README.md
# integration-iterable

[![CircleCI](https://circleci.com/gh/segment-integrations/integration-iterable.svg?style=shield&circle-token=28e826a7a115d0570ef3660a2c2062b0f066f859)](https://circleci.com/gh/segment-integrations/integration-iterable)
  
Iterable server-side integration for [Segment](https://segment.com).


HISTORY.md
1.4.7 / 2017-05-18
==================
  * Add Iterable integration specific mergeNestedObjects option

1.4.6 / 2017-05-15
==================
  * Patch improperly passed category array

1.4.5 / 2017-04-17
==================

  * Bug fixes

1.4.4 / 2017-03-14
==================

  * Patch remaining Date uncaught errors.

1.4.3 / 2017-03-14
==================

  * Standardize Date formats to prevent uncaught exceptions and match Iterable's requirements.

1.4.0 / 2017-01-31
==================

  * Standardize integration (linting, Docker configuration, circle.yml, upgrade
segmentio-integration version, upgrade integration-worker version, etc.)


1.3.2 / 2016-09-06
==================

  * support ecom spec v2

1.3.1 / 2016-08-30
==================

  * map 'phone' to 'phoneNumber'

1.3.0 / 2016-08-16
==================

  * update ecommerce spec syntax to v2
  * include campaignId and templateId for trackPurchase

1.2.3 / 2016-02-16
==================

  * allow calls as long as `userId` or `email`

1.2.2 / 2015-09-02
==================

  * remove revenue->amount alias per iterable request

1.2.1 / 2015-08-27
==================

  * still send call to updateCart api when products is an empty array

1.2.0 / 2015-08-26
==================

  * map ecommerce calls to iterable commerce api

1.1.0 / 2015-04-03
==================

  * remove required email field for identifies

1.0.9 / 2015-04-01
==================

  * mapping page calls to track events

1.0.8 / 2014-12-11
==================

  * rejecting identifies without emails

1.0.7 / 2014-12-08
==================

 * bump segmentio-integration

1.0.6 / 2014-12-03
==================

  * fix: use clone() and traverse()

1.0.5 / 2014-12-02
==================

 * bump integration proto

1.0.4 / 2014-12-02
==================

 * remove .retries()
 * fix dev deps
 * bump dev deps

1.0.3 / 2014-12-02
==================

 * bump segmentio-integration

1.0.2 / 2014-11-25
==================

  * Allow passing of userId instead of email to track calls

1.0.1 / 2014-11-21
==================

 * Bumping segmentio-integration
 * fix build status badge

1.0.0 / 2014-11-14
==================

  * Initial release


test/index.js

var Test = require('segmentio-integration-tester');
var helpers = require('./helpers');
var facade = require('segmentio-facade');
var time = require('unix-time');
var should = require('should');
var assert = require('assert');
var Iterable = require('..');
var objCase = require('obj-case');
var mapper = require('../lib/mapper');
var extend = require('@ndhoule/extend');

describe('Iterable', function(){
  var settings;
  var iterable;
  var payload;
  var test;

  beforeEach(function(){
    payload = {};
    settings = { apiKey: '124175f3654446babf5b5966e232d91d' };
    iterable = new Iterable(settings);
    test = Test(iterable, __dirname);
    test.mapper(mapper);
  });

  it('should have correct settings', function(){
    test
      .name('Iterable')
      .endpoint('https://api.iterable.com/api')
      .channels(['server', 'mobile', 'client'])
      .ensure('settings.apiKey');
  });

  describe('.validate()', function(){
    var msg;

    beforeEach(function(){
      msg = {
        userId: '7331',
        properties: {
          email: 'jd@example.com'
        }
      };
    });

    it('should be invalid when .apiKey is missing', function(){
      delete settings.apiKey;
      test.invalid(msg, settings);
    });

    it('should be valid when .apiKey is given', function(){
      test.valid(msg, settings);
    });

    it('should be invalid when both .userId and .email are missing', function(){
      delete msg.userId;
      delete msg.properties.email;
      test.invalid(msg, settings);
    });

    it('should be valid when .userId is missing but .email is present', function(){
      delete msg.userId;
      test.valid(msg, settings);
    });

    it('should be valid when .userId is present but .email is missing', function(){
      delete msg.properties.email;
      test.valid(msg, settings);
    });
  });

  describe('mapper', function(){
    describe('track', function(){
      it('should map basic track', function(){
        test.maps('track-basic');
      });

      it('should map track with email but no userId', function(){
        test.maps('track-email-no-userId');
      });

      it('should map track with userId but no email', function(){
        test.maps('track-userId-no-email');
      });

      it('should map product added track', function(){
        test.maps('track-added-product');
      });

      it('should map product removed track', function(){
        test.maps('track-removed-product');
      });

      it('should map purchase track', function(){
        test.maps('track-purchase');
      });

      it('should pass empty array for `categories` if track.category() is missing', function(){
        test.maps('track-purchase-no-category');
      });
    });

    describe('identify', function(){
      it('should map basic identify', function(){
        test.maps('identify-basic');
      });

      it('should map basic identify if integration has field other than mergeNestedObjects', function(done){
        var identify = test.fixture('identify-basic');
        identify.input.integrations = {
          Iterable: { somenonspeccedfield: true }
        };
        test
          .set(settings)
          .identify(identify.input)
          .sends(identify.output)
          .pathname('/api/users/update')
          .expects(200, done);
      });

      it('should map basic identify if mergeNestedObjects is set to true', function(done){
        var identify = test.fixture('identify-basic');
        identify.input.integrations = {
          Iterable: { mergeNestedObjects: true }
        };
        test
          .set(settings)
          .identify(identify.input)
          .sends(extend(identify.output, { mergeNestedObjects: true }))
          .pathname('/api/users/update')
          .expects(200, done);
      });

      it('should map basic identify if mergeNestedObjects is set to false', function(done){
        var identify = test.fixture('identify-basic');
        identify.input.integrations = {
          Iterable: { mergeNestedObjects: false }
        };
        test
          .set(settings)
          .identify(identify.input)
          .sends(extend(identify.output, { mergeNestedObjects: false }))
          .pathname('/api/users/update')
          .expects(200, done);
      });

      it('should map a more advanced identify', function(){
        test.maps('identify-advanced');
      });

      it('should map identify with email but no userId', function(){
        test.maps('identify-email-no-userId');
      });

      it('should map identify with userId but no email', function(){
        test.maps('identify-userId-no-email');
      });
    });
  });

  describe('.track()', function(){
    it('should get a good response from the API', function(done){
      var json = test.fixture('track-basic');
      test
        .set(settings)
        .track(json.input)
        .expects(200)
        .pathname('/api/events/track')
        .end(done);
    });

    it('should error on invalid apikey', function(done){
      test
        .set({ apiKey: 'x' })
        .track({})
        .error('Unauthorized', done);
    });

    it('should map Product Added to updateCart if there is a cart provided', function(done){
      var json = test.fixture('track-added-product');

      test
        .set(settings)
        .track(json.input)
        .sends(json.output)
        .expects(200)
        .pathname('/api/commerce/updateCart')
        .end(done);
    });

    it('should not map Product Added to updateCart if there is no cart provided', function(done){
      var json = test.fixture('track-added-product');
      var input = json.input;
      input = objCase.del(input, 'properties.products');

      test
        .set(settings)
        .track(input)
        .expects(200)
        .pathname('/api/events/track')
        .end(done);
    });

    it('should map Product Removed to updateCart if there is a cart provided', function(done){
      var json = test.fixture('track-removed-product');

      test
        .set(settings)
        .track(json.input)
        .sends(json.output)
        .expects(200)
        .pathname('/api/commerce/updateCart')
        .end(done);
    });

    it('should not map Product Removed to updateCart if there is no cart provided', function(done){
      var json = test.fixture('track-removed-product');
      var input = json.input;
      input = objCase.del(input, 'properties.products');

      test
        .set(settings)
        .track(input)
        .expects(200)
        .pathname('/api/events/track')
        .end(done);
    });

    it('should map Order Completed to trackPurchase', function(done){
      var json = test.fixture('track-purchase');

      test
        .set(settings)
        .track(json.input)
        .sends(json.output)
        .expects(200)
        .pathname('/api/commerce/trackPurchase')
        .end(done);
    });

    it('should map a more complex Order Completed to trackPurchase', function(done){
      var json = test.fixture('track-purchase2');

      test
        .set(settings)
        .track(json.input)
        .sends(json.output)
        .expects(200)
        .pathname('/api/commerce/trackPurchase')
        .end(done);
    });

    it('should map a Order Completed with template and campaign ids to trackPurchase', function(done){
      var json = test.fixture('track-purchase-campaign-id');

      test
        .set(settings)
        .track(json.input)
        .sends(json.output)
        .expects(200)
        .pathname('/api/commerce/trackPurchase')
        .end(done);
    });
  });

  describe('.identify()', function(){
    it('should get a good response from the API', function(done){
      test
        .set(settings)
        .identify(helpers.identify())
        .expects(200)
        .end(done);
    });

    it('should error on invalid apikey', function(done){
      test
        .set({ apiKey: 'x' })
        .identify({})
        .error('Unauthorized', done);
    });
  });

  describe('.page()', function(){
    it('should not track automatically', function(done){
      test
        .set(settings)
        .page(helpers.page())
        .requests(0)
        .end(done);
    });

    it('should track all pages', function(done){
      test
        .set(settings)
        .set({ trackAllPages: true })
        .page(helpers.page())
        .requests(1)
        .expects(200)
        .end(done);
    });

    it('should track named pages', function(done){
      test
        .set(settings)
        .set({ trackNamedPages: true })
        .page(helpers.page())
        .requests(1)
        .expects(200)
        .end(done);
    });

    it('should track category pages', function(done){
      test
        .set(settings)
        .set({ trackCategorizedPages: true })
        .page(helpers.page())
        .requests(1)
        .expects(200)
        .end(done);
    });
  });
});


test/helpers.js

var facade = require('segmentio-facade');
var merge = require('merge-util');
var join = require('path').join;
var assert = require('assert');
var uid = require('uid');

/**
 * Create our testing variables
 */

var firstId  = uid();
var secondId = uid();
var groupId  = uid();
var email = 'testing-' + firstId + '@segment.io';
var email2 = 'testing-' + secondId + '@segment.io';

/**
 * Mapper tester.
 *
 * @param {String} dirname
 * @return {Function}
 */

exports.mapper = function(dirname){
  assert(dirname, '__dirname must be supplied');
  dirname = join(dirname, 'fixtures');
  return function(integration){
    integration.fixture = function(name, settings){
      var dir = join(dirname, name + '.json');
      var json = require(dir);
      var input = json.input;
      var output = json.output;
      var type = input.type[0].toUpperCase() + input.type.slice(1);
      var Type = facade[type];
      var map = integration.mapper[input.type];
      var mapped = map.call(integration, new Type(input), settings || {});
      mapped = JSON.parse(JSON.stringify(mapped)); // dates
      mapped.should.eql(output);
    };
  };
};

/**
 * Create ecommerce transaction.
 *
 * @param {Object} options
 * @return {Track}
 */

exports.transaction = function(options){
  return new facade.Track(merge({
    userId: firstId,
    channel: 'server',
    timestamp: new Date,
    event: 'Order Completed',
    properties: {
      orderId: 't-39a224df',
      total: 99.99,
      shipping: 13.99,
      tax: 20.99,
      products: [{
        quantity: 1,
        price: 24.75,
        name: 'Sony Pulse',
        sku: 'p-957c416f',
        category: 'Entertainment'
      }, {
        quantity: 3,
        price: 24.75,
        name: 'Sony PS3',
        sku: 'p-5bd14e17',
        category: 'Entertainment'
      }]
    },
    options: {
      ip: '4.184.68.0',
      userAgent: 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; Trident/6.0)'
    }
  }, options));
};

/**
 * Create a track call merged from `options`
 *
 * @param {Object} options
 * @return {Track}
 */

exports.track = function(options){
  options = options || {};
  return new facade.Track(merge({
    userId: firstId,
    event: 'Baked a cake',
    properties: {
      layers: ['chocolate', 'strawberry', 'fudge'],
      revenue: 19.95,
      numLayers: 10,
      fat: 0.02,
      bacon: '1',
      date: (new Date()).toISOString(),
      address: {
        state: 'CA',
        zip: 94107,
        city: 'San Francisco'
      }
    },
    channel: 'server',
    timestamp: new Date(),
    options: {
      traits: {
        email: options.email || email,
        age: 23,
        created: new Date(),
        bad: null,
        alsoBad: undefined,
        address: {
          state: 'CA',
          zip: 94107,
          city: 'San Francisco'
        }
      },
      ip: '4.184.68.0',
      userAgent: 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; Trident/6.0)'
    }
  }, options));
};

/**
 * Create a bare track call merged from `options`
 *
 * @param {Object} options
 * @return {Track}
 */

exports.track.bare = function(options){
  return new facade.Track(merge({
    userId: 'aaa',
    event: 'Bear tracks',
    channel: 'server'
  }, options || {}));
};

/**
 * Create an identify call merged from `options`
 *
 * @param {Object} options
 * @return {Identify}
 */

exports.identify = function(options){
  options = options || {};
  return new facade.Identify(merge({
    userId: firstId,
    traits: {
      fat: 0.02,
      firstName: 'John',
      'Last Name': 'Doe',
      email: options.email || email,
      company: 'Segment.io',
      city: 'San Francisco',
      state: 'CA',
      phone: '5555555555',
      websites: [
        'http://calv.info',
        'http://ianstormtaylor.com',
        'http://ivolo.me',
        'http://rein.pk'
      ],
      bad: null,
      alsoBad: undefined,
      met: (new Date()).toISOString(),
      created: new Date('1/12/2013'),
      userAgent: 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; Trident/6.0)'
    },
    context: {
      ip: '12.212.12.49'
    },
    timestamp: new Date(),
    channel: 'server'
  }, options));
};

/**
 * Create a page call merged from `options`
 *
 * @param {Object} options
 * @return {Page}
 */

exports.page = function(options){
  return new facade.Page(merge({
    userId: firstId,
    name: 'Docs',
    category: 'Support',
    properties: {
      url: 'https://segment.io/docs',
      title: 'Analytics.js - Segment.io',
      email: email
    },
    context: {
      ip: '12.212.12.49'
    },
    timestamp: new Date,
    channel: 'server'
  }, options || {}));
};

/**
 * Create a screen call merged from `options`
 *
 * @param {Object} options
 * @return {Page}
 */

exports.screen = function(options){
  return new facade.Screen(merge({
    userId: firstId,
    name: 'Login',
    category: 'Authentication',
    properties: {
      type: 'Facebook'
    },
    context: {
      ip: '12.212.12.49'
    },
    timestamp: new Date,
    channel: 'server'
  }, options || {}));
};

/**
 * Create a group call merged from `options`
 *
 * @param {Object} options
 * @return {Group}
 */

exports.group = function(options){
  return new facade.Group(merge({
    groupId: groupId,
    userId: firstId,
    traits: {
      email: email,
      name: 'Segment.io',
      state: 'CA',
      city: 'San Francisco',
      created: new Date('2/1/2014'),
      plan: 'Enterprise',
    },
    context: {
      ip: '12.212.12.49'
    },
    timestamp: new Date,
    channel: 'server'
  }, options || {}));
};

/**
 * Create an alias call merged from `options`
 *
 * @param {Object} options
 * @return {Alias}
 */

exports.alias = function(options){
  return new facade.Alias(merge({
    from: email,
    to: email2,
    channel: 'server',
    timestamp: new Date()
  }, options || {}));
};


lib/index.js

/**
 * Module dependencies.
 */

var integration = require('segmentio-integration');
var type = require('type-component');
var mapper = require('./mapper');
var get = require('obj-case');
var isEmail = require('is-email');

/**
 * Expose `Iterable`
 */

var Iterable = module.exports = integration('Iterable')
  .endpoint('https://api.iterable.com/api')
  .channels(['server', 'mobile', 'client'])
  .ensure('settings.apiKey')
  .mapToTrack(['page'])
  .mapper(mapper)
  .retries(2);

/**
 * We require either a userId or an email
 */

Iterable.ensure(function(msg, settings){
  if (msg.userId() || msg.email()) return;
  return this.reject('Must set either userId or email');
});

/**
 * We require both IDs be emails for alias to work in iterable
 */

Iterable.ensure(function(msg, settings){
  if (msg.action() !== 'alias') return;
  if (isEmail(msg.previousId()) && isEmail(msg.userId())) return;
  return this.reject('Alias calls only work for emails in iterable');
});

/**
 * Identify.
 *
 * https://api.iterable.com/api/docs#!/users/updateUser_post_1
 *
 * @param {Object} payload
 * @param {Function} fn
 * @api public
 */

Iterable.prototype.identify = function(identify, fn){
  return this
    .post('/users/update')
    .type('json')
    .set(headers(this.settings))
    .send(identify)
    .end(this.handle(fn));
};

/**
 * Track.
 *
 * https://api.iterable.com/api/docs#!/events/track_post_0
 *
 * @param {Track} track
 * @param {Function} fn
 * @api public
 */

Iterable.prototype.track = function(track, fn){
  return this
    .post('/events/track')
    .type('json')
    .set(headers(this.settings))
    .send(track)
    .end(this.handle(fn));
};

/**
 * Alias.
 *
 * https://api.iterable.com/api/docs#!/users/updateEmail_post_1
 *
 * @param {Alias} alias
 * @param {Function} fn
 * @api public
 */

Iterable.prototype.alias = function(alias, fn) {
  console.log(alias);
  return this
    .post('/users/updateEmail')
    .type('json')
    .set(headers(this.settings))
    .send(alias)
    .end(this.handle(fn));
};

/**
 * Product Added/Removed.
 *
 * Regular track call, unless properties.products is present
 *
 * @param {Track} track
 * @param {Function} fn
 * @api public
 */

Iterable.prototype.productAdded =
Iterable.prototype.productRemoved = function(track, fn){
  var products = get(track.properties(), 'products');
  if (type(products) === 'array') {
    return this
      .post('/commerce/updateCart')
      .type('json')
      .set(headers(this.settings))
      .send(mapper.productAdded(track))
      .end(this.handle(fn));
  } else {
    return this
      .post('/events/track')
      .type('json')
      .set(headers(this.settings))
      .send(mapper.track(track))
      .end(this.handle(fn));
  }
};

/**
 * Order Completed.
 *
 * https://api.iterable.com/api/docs#!/events/track_post_0
 *
 * @param {Track} track
 * @param {Function} fn
 * @api public
 */

Iterable.prototype.orderCompleted = function(track, fn){
  return this
    .post('/commerce/trackPurchase')
    .type('json')
    .set(headers(this.settings))
    .send(mapper.orderCompleted(track))
    .end(this.handle(fn));
};

function headers(settings) {
  return {
    'User-Agent': 'Segment.io/1.0',
    'Api-Key': settings.apiKey
  };
}


lib/mapper.js
/**
 * Module dependencies.
 */

var Track = require('segmentio-facade').Track;
var traverse = require('isodate-traverse');
var foldl = require('@ndhoule/foldl');
var unixTime = require('unix-time');
var del = require('obj-case').del;
var reject = require('reject');
var pick = require('@ndhoule/pick');
var extend = require('@ndhoule/extend');
var includes = require('@ndhoule/includes');
var is = require('is');

/**
 * Map `identify`.
 *
 * @param {Identify} identify
 * @return {Object}
 * @api private
 */

exports.identify = function(identify) {
  // "phoneNumber" field in Iterable should be mapped from "phone" in Segment spec
  var traits = formatDates(extend(identify.traits({
    phone: 'phoneNumber',
    timestamp: 'profileUpdatedAt'
  }), { met: identify.created() }));

  del(traits, 'id');

  var context = extend(identify.context(), { timeZone: identify.timezone() });
  var whitelist = ['app', 'device', 'ip', 'locale', 'location', 'page', 'timeZone', 'userAgent'];
  context = formatDates(pick(whitelist, context));

  var payload = {
    email: identify.email(),
    userId: identify.userId(),
    dataFields: reject(extend(context, traits))
  };

  var iterableIntegrationFields = pick(['mergeNestedObjects'], identify.options('Iterable'));
  return extend(payload, iterableIntegrationFields);
};

/**
 * Map `track`.
 *
 * @param {Track} track
 * @return {Object}
 * @api private
 */

exports.track = function(track){
  var properties = formatDates(track.properties());

  return {
    email: track.email(),
    userId: track.userId(),
    eventName: track.event(),
    createdAt: unixTime(track.timestamp()),
    dataFields: properties
  };
};

/**
 * Map `Product Added/Removed`.
 *
 * This is only invoked when a full products array is present
 *
 * @param {Track} track
 * @return {Object}
 * @api private
 */

exports.productAdded =
exports.productRemoved = function(track){
  return {
    user: {
      email: track.email(),
      userId: track.userId()
    },
    items: formatProducts(track.products())
  };
};

/**
 * Map `Order Completed`.
 *
 * @param {Track} track
 * @return {Object}
 * @api private
 */

exports.orderCompleted = function(track){
  var specialFields = ['templateId', 'campaignId'];
  var props = track.properties();
  var result = {
    createdAt: unixTime(track.timestamp()),
    total: track.total(),
    user: {
      userId: track.userId(),
      email: track.email()
    },
    dataFields: foldl(function(acc, val, key){
      if (key !== 'total' && key !== 'products' && !includes(key, specialFields)) acc[key] = val;
      return acc;
    }, {}, props),
    items: formatProducts(track.products())
  };
  extend(result, pick(specialFields, props));
  return result;
};

/**
 * Turn a Segment products array into Iterable format
 * Specifically, any non-special fields should be nested
 * under a field called "dataFields"
 *
 * @param products
 * @returns {Array}
 */

function formatProducts(products){
  return products.map(function(item){
    var product = new Track({ properties: item });
    var category = product.category();
    var payload = reject({
      id: product.productId() || product.id(),
      sku: product.sku(),
      name: product.name(),
      categories: category ? [category] : [],
      price: product.price(),
      quantity: product.quantity(),
      description: product.proxy('properties.description'),
      imageUrl: product.proxy('properties.imageUrl'),
      url: product.proxy('properties.url')
    });

    payload.dataFields = foldl(function(acc, val, key){
      if (!payload.hasOwnProperty(key) && key !== 'category' && key !== 'product_id') acc[key] = val;
      return acc;
    }, {}, product.properties());

    return payload;
  });
}

/**
 * convert a date into the standard Iterable format "yyyy-MM-dd HH:mm:ss ZZ" (for example, 2015-04-15 10:46:45 +00:00), by fiddling with the ISO string (i.e., 2015-04-15T10:46:45.808Z)
 * @param date
 * @returns {string}
 **/

function dateToIterableDateStringFormat(date) {
  if (typeof date.getDate === 'function' && !Number.isNaN(date.getDate())) return date.toISOString().replace('T', ' ').split('.')[0] + " +00:00";
}

/**
* Format Dates.
*
* https://support.iterable.com/hc/en-us/articles/208183076-Data-Field-Types-in-Iterable
*
* @param {Object} obj Traits or properties
* @return {Object}
**/

function formatDates(obj) {
  return foldl(function(acc, val, key) {
    if (is.date(val)) {
      acc[key] = dateToIterableDateStringFormat(val);
    } else {
      acc[key] = val;
    }
    return acc;
  }, {}, traverse(obj));
}
